File: Data Structures & Algorithms/search-2d-matrix/submission-9.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        //int[][] matrix = new int[4][5];
        //int[][] matrix = new int{{1,2,3}, {4,5,6}}
        int m = matrix.length;
        int n = matrix[0].length;

        int low=0;
        int high=m*n-1;

        while(low <= high){
            int mid = low + (high-low)/2;
            int r = mid / n;
            int c = mid % n;

            if(matrix[r][c] == target){
                return true;
            }
            else if(matrix[r][c] < target){
                low = mid + 1;
            }
            else{
                high = mid-1;
            }
        }

        return false;
        
    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-9.py
class Solution:
   #rotated sorted creates two sorted poritons LSP and RSP
   #we are LSP if nums[L] <= nums[mid], else RSP
   #do a binary search, but if we are not in correct portion, go to other
   def search(self, nums: List[int], target: int) -> int:
       L = 0
       R = len(nums) - 1

       while L <= R:
         #compute mid index and check if its target
         mid = L + (R - L)//2
         if nums[mid] == target:
            return mid

         #LSP
         if nums[L] <= nums[mid]:
            #we are in correct portion     
            if nums[L] <= target < nums[mid]:
               R = mid - 1
            #otherwise, go right
            else:
               L = mid + 1          
         #RSP
         else:
            #we are in correct portion
            if nums[mid] < target <= nums[R]:
               L =  mid + 1
            #otherwise go left
            else:
               R = mid -1

       return -1
        